### PR TITLE
Add embedded find/replace dialog to editor widget

### LINKED ISSUE #24531: Workbench: Add search/replace functionality to editor
MantidPlot's script editor has a search and replace mechanism. We need something similar in workbench.

Rather than a dialog can we consider a toolbar that pops up at the top of the editor like in PyCharm?
py_replace_with_regex

The same options that are available in MantidPlot's version should be available at a minimum.

### PR DESCRIPTION
**Pull request by dtasev**
This PR is waiting on ~~#24700~~ to reuse some of the memory leaking common functionality.

**Description of work.**
Adds embedded find/replace dialog into the editor widget. To trigger press `Ctrl+F`.

Also fixes the editor widgets (and children) being leaked.

**To test:**
- Press CTRL+F in the editor. Do a bunch of search/replaces. Try to check if all the options work correctly. Try some unicode symbols: `áéóúí`.
- Test that search is performed independently for each editor
- Test pressing escape to close the find/replace, only works if Find/Replace has focus (seems to be `QsciScintilla` limitation - it doesn't allow `Escape` trigger through the editor itself, I am not sure why, it doesn't seem to be used for anything)
- Added `Enter` to search next, `Shift+Enter` to search previous, only works if Find/Replace has focus

Fixes #24531 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DTasev** [comment]
Note to self: AssertionError: Todo after #24700 has been merged, to reuse the helper functions.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is really nice. I much prefer this to the dialog box from MantidPlot. It looks like it works entirely as I would expect :+1:

As we discussed offline there are a few cosmetic things to look into:
* [x] we need an item in the menu so that users know the shortcut
* [x] can we have next/previous types buttons instead of search backwards?
* [x] can we shrink the borders as whole bar seems quite visually imposing
* [x] replace/replace all buttons moved to bottom next to replace box
* [x] is it possible to prefil the search box if I have a word hightlighted? PyCharm does this and it saves the copy step

**Comment by DTasev** [comment]
I have pushed up changes to address the comments. Note that I've updated the test instructions in the main post, most notably enter/shift+enter searches next/previous, escape closes the find/replace. The layout is a lot tighter, but I can't seem to find a way to stop the spacers from expanding when the window is made very wide.

**Comment by DTasev** [comment]
I figured it out - replaced the horizontal spacer with a disabled flat button that _should_ be invisible. (it's above Replace All)
image

There is still empty space there, above Replace All, but I am not sure what makes sense to go there, we don't have enough options for buttons.

**Comment by martyngigg** [comment]
@DTasev It looks like MultiFileInterpreter test is failing

**Comment by DTasev** [comment]
It should be fixed now

**Review by martyngigg — CHANGES_REQUESTED** [review]
The functionality seems to work for me with the exception of "match case", for example typing "IMPORT" into the search still matches "import".

I also have an issue on Ubuntu with words being cut off:
screenshot from 2019-02-14 13-54-10

I guess this is a theme problem with fixed button sizes?

**Comment by DTasev** [comment]
> The functionality seems to work for me with the exception of "match case", for example typing "IMPORT" into the search still matches "import".

That might be a case of the combo box instantly replacing "IMPORT" back with "import" as the lower-case version was in the history and then the search is performed with lower-case again. This is the default behaviour of QComboBox, and it matches MantidPlot which is why I haven't changed it.

> I also have an issue on Ubuntu with words being cut off:
> screenshot from 2019-02-14 13-54-10
> 
> I guess this is a theme problem with fixed button sizes?

Yeah, I will make the maximum widths more relaxed, and that should fix it

File: Framework/PythonInterface/test/python/mantid/plots/helperfunctionsTest.py [hunk 1]
```diff
@@ -526,21 +526,6 @@ def test_get_bins_with_dy(self, ws):
         self.assertTrue(np.array_equal([2.0, 5.0, 8.0, 11.0], y))
         self.assertTrue(np.array_equal([2.0, 5.0, 8.0, 11.0], dy))
 
-    @add_md_workspace_with_data()
```
Comment (martyngigg) [inline Framework/PythonInterface/test/python/mantid/plots/helperfunctionsTest.py]: This just removes a duplicate test.

**Review by martyngigg — DISMISSED** [review]

**Comment by martyngigg** [comment]
It looks like we still have some failing tests.

**Comment by DTasev** [comment]
Tests are now passing!

**Review by martyngigg — APPROVED** [review]

**Pull request merged**